### ISSUE TITLE
Hazelcast .NET Client v5.0 compatible with v4.2 server? [API-1218]

### ISSUE THREAD

**Issue by drieseng** [hazelcast/hazelcast-csharp-client#579]
Can you clarify whether the V5.0 client ought to be compatible with a V4.2 IMDG?
When we attempt to establish a connect using the code fragment below it just blocks indefinitly:

```cs
using System;
using System.Threading.Tasks;
using Hazelcast;
using Hazelcast.Networking;

namespace Hazelcastv5
{
    class Program
    {
        public static async Task Main()
        {
            var options = new HazelcastOptionsBuilder().With((c, o) =>
                {
                    o.ClusterName = "dev";
                    o.Networking.Addresses.Add("xxx.xx.xx.xxx:5701");
                    o.Networking.ConnectionTimeoutMilliseconds = 1000;
                    o.Networking.ReconnectMode = ReconnectMode.DoNotReconnect;
                    o.Networking.SmartRouting = false;
                }).Build();

            var client = await HazelcastClientFactory.StartNewClientAsync(options);

            Console.WriteLine($"Connection established to cluster '{client.ClusterName}'");
        }
    }
}
```

This is what we see in our server logs:

```txt
2022-02-03 09:59:57,957 [ INFO] [hz.determined_jepsen.priority-generic-operation.thread-0] [c.h.c.i.p.t.AuthenticationMessageTask]: [xxx.xx.xx.xxx]:5701 [dev] [4.2] Received auth from Connection[id=2176747, /xxx.xx.xx.xxx:5701->/xxx.xx.xx.xxx:59287, qualifier=null, endpoint=[xxx.xx.xx.xxx]:59287, alive=true, connectionType=CSP, planeIndex=-1], successfully authenticated, clientUuid: 564a48ae-5fff-e4d8-c4f0-08e075190c92, client version: 5.0
2022-02-03 10:00:02,719 [ INFO] [hz.determined_jepsen.IO.thread-in-0] [c.h.i.s.t.TcpServerConnection]: [xxx.xx.xx.xxx]:5701 [dev] [4.2] Connection[id=2176747, /xxx.xx.xx.xxx:5701->/xxx.xx.xx.xxx:59287, qualifier=null, endpoint=[xxx.xx.xx.xxx]:59287, alive=false, connectionType=CSP, planeIndex=-1] closed. Reason: Connection closed by the other side
2022-02-03 10:00:07,883 [ INFO] [hz.determined_jepsen.cached.thread-15] [c.h.c.i.ClientEndpointManager]: [xxx.xx.xx.xxx]:5701 [dev] [4.2] Destroying ClientEndpoint{connection=Connection[id=2176747, /xxx.xx.xx.xxx:5701->/xxx.xx.xx.xxx:59287, qualifier=null, endpoint=[xxx.xx.xx.xxx]:59287, alive=false, connectionType=CSP, planeIndex=-1], clientUuid='564a48ae-5fff-e4d8-c4f0-08e075190c92, authenticated=true, clientVersion=5.0, creationTime=1643882397957, latest clientAttributes=null, labels=[]}```

**Comment by zpqrtbnk** [comment]
Client 5.0 should connect to server 4.2, yes. Surprised this simple test is failing. Investigating.

**Comment by zpqrtbnk** [comment]
Client 5.0 happily connects to server 5.x and 4.x as long as you comment out the `o.Networking.SmartRouting = false` bit. It fails to connect to server 5.x and 4.x if you leave it uncommented. That would indicate a regression in 5.0 with regards to non-smart routing + a lack of tests, as we should have detected it. Investigating.

**Comment by drieseng** [comment]
@zpqrtbnk We did some more tests, and we cannot even get the 5.0 client to work against a 5.0.2 server. Neither with nor without smart routing.

|Client|Server|Result|
|------|-------|-------|
|4.0.1<br>**SmartRouting**=**false**|4.2|OK|
|5.0.0<br>**SmartRouting**=**false**|4.2|NOT OK: Hangs on client creation:<br><br>info: Hazelcast.Clustering.Heartbeat[0]<br>Heartbeat with 5000ms period and 60000ms timeout<br>info: Hazelcast.Clustering.ClusterConnections[0]<br>Authenticated client 'XXX' (b37f33d) running version 5.0.0+eac6f8 on connection 7c93747 from XXX.XX.XX.XXX:50545 to member 564d650 at XXX.XX.XX.XXX:5701 of cluster 'dev' (34440da) running version 4.2.<br>info: Hazelcast.Clustering.ClusterMembers[0]<br>      Members [1] {<br>          564d650 - XXX.XX.XX.XXX:5701* / null - new<br>      }<br>warn: Hazelcast.Clustering.ClusterEvents[0]<br>      No event handler for [2]<br>warn: Hazelcast.Clustering.MemberConnection[0]<br>      Received message for unknown invocation 7c93747:2.
|5.0.0<br>**SmartRouting**=**true**|4.2|NOT OK: Hangs on client creation:<br><br>info: Hazelcast.Clustering.Heartbeat[0]<br>    Heartbeat with 5000ms period and 60000ms timeout<br>info: Hazelcast.Clustering.ClusterConnections[0]<br>Authenticated client 'XXX' (56830ea) running version 5.0.0+eac6f8 on connection b776444 from XXX.XX.XX.XXX:54552 to member af4b6bc at XXX.XX.XX.XXX:5701 of cluster 'dev' (c24bceb) running version 4.2.<br>info: Hazelcast.Clustering.ClusterMembers[0]<br>      Members [1] {<br>          af4b6bc - XXX.XX.XX.XXX:5701* / null - new<br>      }<br>warn: Hazelcast.Clustering.MemberConnection[0]<br>      Received message for unknown invocation b776444:2.|
|5.0.0<br>**SmartRouting**=**false**|5.0.2|NOT OK: Hangs on client creation:<br><br>info: Hazelcast.Clustering.Heartbeat[0]<br>Heartbeat with 5000ms period and 60000ms timeout<br>info: Hazelcast.Clustering.ClusterConnections[0]<br>Authenticated client 'XXX' (fa4d1c9) running version 5.0.0+eac6f8 on connection e3a89fe from XXX.XX.XX.XXX:54157 to member 8c446fe at XXX.XX.XX.XXX:5701 of cluster 'dev' (8a455b7) running version 5.0.2.<br>info: Hazelcast.Clustering.ClusterMembers[0]<br>Members [1] {<br>8c446fe - XXX.XX.XX.XXX:5701* / null - new<br>      }<br>warn: Hazelcast.Clustering.MemberConnection[0]<br>      Received message for unknown invocation e3a89fe:2.
|5.0.0<br>**SmartRouting**=**true**|5.0.2|NOT OK: Hangs on client creation:<br><br>info: Hazelcast.Clustering.Heartbeat[0]<br>Heartbeat with 5000ms period and 60000ms timeout<br>info: Hazelcast.Clustering.ClusterConnections[0]<br>Authenticated client 'XXX' (1868890) running version 5.0.0+eac6f8 on connection fab5d2c from XXX.XX.XX.XXXX:49870 to member 394f1ef at XXX.XX.XX.XXX:5701 of cluster 'dev' (6e40ad2) running version 5.0.2.<br>info: Hazelcast.Clustering.ClusterMembers[0]<br>      Members [1] {<br>          394f1ef - XXX.XX.XX.XXX:5701* / null - new<br>      }<br>warn: Hazelcast.Clustering.MemberConnection[0]<br>      Received message for unknown invocation fab5d2c:2.

We used the following options:
```cs
var options = new HazelcastOptionsBuilder().With((_, o) => {
    o.ClusterName = "dev";
    o.Networking.ReconnectMode = ReconnectMode.DoNotReconnect;
    o.Networking.Addresses.Add("XXX.XX.XX.XXX:5701");
    o.Networking.ConnectionTimeoutMilliseconds = (int) TimeSpan.FromMinutes(1).TotalMilliseconds;
    o.Networking.SmartRouting = false; // or true
    o.LoggerFactory.Creator = () => Microsoft.Extensions.Logging.LoggerFactory.Create(builder => builder.AddConsole());
   }),
                                          .With("Logging:LogLevel:Default", "Debug")
                                          .With("Logging:LogLevel:System", "Debug")
                                          .With("Logging:LogLevel:Microsoft", "Debug")
                                          .Build();
```
Is it a known issue that the client can hang indefinitely when the connection cannot be established?

**Comment by zpqrtbnk** [comment]
Re. the last question: for some reasons, the connection timeout is infinite, and every failure leads to a retry, so yes, the client ends up hanging. Would it make sense to have a default timeout? Yes. Would this be a breaking change that we are not ready to do now? Yes.

As for the rest, I am... super annoyed you are facing these issues that should not be issues. Running more tests today.

**Comment by drieseng** [comment]
@zpqrtbnk If you look at the code snippet that we use configure the options, you'll see that we:
* set the connection timeout to one minute.
   we tried setting it to a much lower value, but that didn't change anything.
* set the reconnect mode to *DoNotReconnect*.

I don't see why the client would hang indefinitely. I'll also try to do more tests tomorrow.

**Comment by drieseng** [comment]
@zpqrtbnk Will this fix be included in v5.1? Any idea when we can expect this?

**Comment by emreyigit** [comment]
@drieseng Yes, it is in v5.1, and the release process has started for v5.1. Assuming that will be released next week most lately.
